docs/devel/e2e-tests.md

```diff
@@ -232,6 +232,102 @@ stale permissions can cause problems.
 
   - `sudo iptables -F`, clear ip tables rules left by the kube-proxy.
 
+### Federation e2e tests
+
+By default, `e2e.go` provisions a single Kubernetes cluster, and any `Feature:Federation` ginkgo tests will be skipped.
+
+Federation e2e testing involve bringing up multiple "underlying" Kubernetes clusters,
+and deploying the federation control plane as a Kubernetes application on the underlying clusters.
+
+The federation e2e tests are still managed via `e2e.go`, but require some extra configuration items.
+
+#### Configuring federation e2e tests
+
+The following environment variables will enable federation e2e building, provisioning and testing.
+
+```sh
+$ export FEDERATION=true
+$ export E2E_ZONES="us-central1-a us-central1-b us-central1-f"
+```
+
+A Kubernetes cluster will be provisioned in each zone listed in `E2E_ZONES`. A zone can only appear once in the `E2E_ZONES` list.
+
+#### Image Push Repository
+
+Next, specify the docker repository where your ci images will be pushed.
+
+If you are using `KUBERNETES_PROVIDER=gce`, you can simply set your push repo base based on your project name, and the
+necessary repositories will be auto-created when you first push your container images:
+
+```sh
+$ export FEDERATION_PUSH_REPO_BASE="gcr.io/${GCE_PROJECT_NAME}"
+```
+
+Otherwise, if the provider is not gce, you'll be responsible for creating and managing access to the repositories manually.
+
+```sh
+$ export FEDERATION_PUSH_REPO_BASE="quay.io/colin_hom"
+```
+
+Given this example, the `federation-apiserver` container image will be pushed to the repository `quay.io/colin_hom/federation-apiserver`.
+
+The docker client on the machine running `e2e.go` must have push access for the following pre-existing repositories:
+
+* `${FEDERATION_PUSH_REPO_BASE}/federation-apiserver`
+* `${FEDERATION_PUSH_REPO_BASE}/federation-controller-manager`
+
+These repositories must allow public read access, as the e2e node docker daemons will not have any credentials.
+
+#### Build
+* Compile the binaries and build container images:
+
+  ```sh
+  $ KUBE_RELEASE_RUN_TESTS=n KUBE_FASTBUILD=true go run hack/e2e.go -v -build
+  ```
+
+* Push the federation container images
+
+  ```sh
+  $ build/push-federation-images.sh
```

Comment:
`e2e.go -push` exists, and is broken and [intentionally disabled](https://github.com/kubernetes/kubernetes/blob/master/cluster/kube-push.sh#L26) - one of the main reasons pushing federation images exists as separate step.

Unfortunately, it can not be part of -build either- the jenkins e2e runners expect `-build` and `push-ci-build.sh` to be separate steps, and as this is a gcloud-authenticated action against your gce project- image pushing is part of pushing the ci build.